vassal-app/src/main/java/VASSAL/counters/MatCargo.java

```diff
@@ -217,46 +233,63 @@ public Command makeSetMatCommand(GamePiece newMat) {
   public Command findNewMat(Map map, Point pt) {
     Command comm = new NullCommand();
     if (map != null) {
-      GamePiece newMat = map.findAnyPiece(pt, PieceFinder.MAT_ONLY);
+      final GamePiece newMat = locateNewMap(map, pt);
 
-      if ((newMat == null) && (detectionDistanceX != 0) || (detectionDistanceY != 0)) {
-        final Point pt2 = new Point();
+      if (newMat != null) {
+        final Mat mat = (Mat) Decorator.getDecorator(newMat, Mat.class);
+        if (mat != null) {
+          comm = comm.append(mat.makeAddCargoCommand(getOutermost(this)));
+        }
+      }
+      else {
+        // We're NOT on a mat, so if we WERE on one, mark ourselves as no longer on it
+        comm = comm.append(makeClearMatCommand());
+      }
+    }
+    return comm;
+  }
 
-        pt2.x = pt.x + detectionDistanceX;
+  /**
+   * Non Command generating part of findNewMat(). Used to check for a potential
+   * valid mat at a target location without generating the Mat/Cargo commands
+   * @param map map to check
+   * @param pt point to check
+   * @return the Mat GamePiece at map.point or null if none
+   */
+  public GamePiece locateNewMap(Map map, Point pt) {
```

Comment:
I'm slightly confused by the method name -- it is called locateNew **_Map_** but it seems to be finding and returning a **_Mat_**?